pkg/storage/tsdb/index/shard.go

```diff
@@ -15,24 +15,50 @@ const (
 )
 
 // ShardAnnotation is a convenience struct which holds data from a parsed shard label
+// Of MUST be a power of 2 to ensure sharding logic works correctly.
 type ShardAnnotation struct {
 	Shard uint32
 	Of    uint32
 }
 
+func NewShard(x, of uint32) ShardAnnotation {
+	return ShardAnnotation{
+		Shard: x,
+		Of:    of,
+	}
+}
+
+// Match returns whether a fingerprint belongs to a certain shard.
+// The Shard must be a power of 2.
+// Inclusion in a shard is calculated by determining the arbitrary bit prefix
+// for a shard, then ensuring the fingerprint has the same prefix
 func (shard ShardAnnotation) Match(fp model.Fingerprint) bool {
-	return uint64(fp)%uint64(shard.Of) == uint64(shard.Shard)
+	requiredBits := shard.RequiredBits()
+
+	// A shard only matches a fingerprint when they both start with the same prefix
```

Comment:
How does it compare to modulo with regards to the balancing of series ? Can we add a test that creates millions of fake series and check how both Matches method balance series ?

I really love the PR, I'm just cautious about this because now that the method is hardcoded in tsdb it won't be easily changed.

I think we're good but for posterity I'd like to have that tests so we can trust our code in term of balancing.

Reply:
I relied on the uniformity principle of hash functions for this (evenly distributed over bits), but it's a really good call out to test this. I wrote a quick script to analyze an index from one of our clusters and got the following results (old sharding logic on the left, new logic on the right):

```
Shard: 0  => 6538 vs 6614

Shard: 1  => 6453 vs 6498

Shard: 2  => 6630 vs 6408

Shard: 3  => 6516 vs 6475

Shard: 4  => 6488 vs 6598

Shard: 5  => 6474 vs 6457

Shard: 6  => 6526 vs 6662

Shard: 7  => 6662 vs 6631

Shard: 8  => 6660 vs 6534

Shard: 9  => 6572 vs 6554

Shard: 10  => 6497 vs 6552

Shard: 11  => 6580 vs 6641

Shard: 12  => 6570 vs 6383

Shard: 13  => 6503 vs 6605

Shard: 14  => 6473 vs 6492

Shard: 15  => 6439 vs 6477
```

tl;dr: This is very similar and I'm comfortable with the new calculations.